message.go

```diff
@@ -4,54 +4,246 @@ import (
-	err             interface{}
-	remainingFields int
-	decodeSlots     []interface{}
-}
-
-func decodeIntoMessage(dec decoder, m *message) error {
-	for _, s := range m.decodeSlots {
-		if err := decodeMessage(dec, m, s); err != nil {
-			return err
+type RPCMessage interface {
+	Type() MethodType
+	Decode(l int, d decoder, p *protocolHandler, cc *callContainer) error
+	RPCData
+}
+
+type RPCData interface {
+	Name() string
+	Arg() interface{}
+	SeqNo() seqNumber
+	Err() error
+	Res() interface{}
+	ResponseCh() chan RPCMessage
+	MinLength() int
+	DecodeData(int, decoder, *protocolHandler, *callContainer) error
+}
+
+type BasicRPCData struct{}
+
+func (BasicRPCData) Name() string {
+	panic("Name() not implemented for this type")
+}
+
+func (BasicRPCData) Arg() interface{} {
+	panic("Arg() not implemented for this type")
+}
+
+func (BasicRPCData) SeqNo() seqNumber {
+	panic("SeqNo() not implemented for this type")
+}
+
+func (BasicRPCData) Err() error {
+	panic("Err() not implemented for this type")
+}
+
+func (BasicRPCData) Res() interface{} {
+	panic("Res() not implemented for this type")
+}
+
+func (BasicRPCData) ResponseCh() chan RPCMessage {
+	panic("ResponseCh() not implemented for this type")
+}
+
+type RPCCallData struct {
+	BasicRPCData
+	seqno seqNumber
+	name  string
+	arg   interface{}
+}
+
+func (RPCCallData) MinLength() int {
+	return 3
+}
+
+func (r *RPCCallData) DecodeData(l int, d decoder, p *protocolHandler, _ *callContainer) (err error) {
+	if err = d.Decode(&r.seqno); err != nil {
+		return err
+	}
+	if err = d.Decode(&r.name); err != nil {
+		return err
+	}
+	if r.arg, err = p.getArg(r.name); err != nil {
+		return err
+	}
+	return d.Decode(r.arg)
+}
+
+func (r RPCCallData) SeqNo() seqNumber {
+	return r.seqno
+}
+
+func (r RPCCallData) Name() string {
+	return r.name
+}
+
+func (r RPCCallData) Arg() interface{} {
+	return r.arg
+}
+
+type RPCResponseData struct {
+	BasicRPCData
+	c   *call
+	err error
+}
+
+func (r RPCResponseData) MinLength() int {
+	return 3
+}
+
+func (r *RPCResponseData) DecodeData(l int, d decoder, _ *protocolHandler, cc *callContainer) error {
+	var seqNo seqNumber
+	if err := d.Decode(&seqNo); err != nil {
+		return err
+	}
+
+	// Attempt to retrieve the call
+	r.c = cc.RetrieveCall(seqNo)
+	if r.c == nil {
+		return CallNotFoundError{seqNo}
+	}
+
+	// Decode the error
+	var responseErr interface{}
+	if r.c.errorUnwrapper != nil {
+		responseErr = r.c.errorUnwrapper.MakeArg()
+	}
+	if err := d.Decode(responseErr); err != nil {
+		return err
+	}
+
+	// Ensure the error is wrapped correctly
+	if r.c.errorUnwrapper != nil {
+		var dispatchErr error
+		r.err, dispatchErr = r.c.errorUnwrapper.UnwrapError(responseErr)
+		if dispatchErr != nil {
+			return dispatchErr
+		}
+	} else if responseErr != nil {
+		errAsString, ok := responseErr.(*string)
+		if !ok {
+			return errors.New("unable to convert error to string")
 		}
+		r.err = errors.New(*errAsString)
+	}
+
+	// Decode the result
+	if r.c.res == nil {
```

Comment:
when is a nil value for .res allowed? I'd prefer to just let it panic in that case, instead of hiding the error, unless there's some existing use case.